### ISSUE TITLE
Some general question regarding the numbers.

### ISSUE THREAD

**Issue by AdamKHacklaender** [explosion/spaCy#5537]
Hi!

I am quite new to NLP and especially spaCy, and I have some questions that I didn't seem to find a clear answer for:

1) What loss function is used by spaCy? I've read that it's a loss function but I am certain it's more complicated than that.

2) How does one determine the appropriate values for n of iterations, compounding factor, and dropout rate? 

3) When training a blank model with 4-5 entities, is it better to train them at once or start with just one? What about the length of sentences/paragraphs to train on? 

4) Is it a common practice to train over and over again on one given model or, rather, train one time once all data is collected? Is consequent training (or even consequent training on the same training data) detrimental to the model?

I am aware those are beginner questions and I would be more than happy for any answer or a direction as to what concepts to get familiar with (the latter is better).

Thanks :) 

## Your Environment

* Operating System: MacOS High Sierra
* Python Version Used: 3.7.4
* spaCy Version Used: 2.2.4

**Comment by honnibal** [comment]
Hi @adamhadani ,

Sorry for the delay getting back to you on this. There aren't straight-forward answers to all of your questions, so I hope the below is useful.

> What loss function is used by spaCy? I've read that it's a loss function but I am certain it's more complicated than that.

This depends on the individual components. Most of the built-in components use some sort of categorical cross-entropy loss, although the text categorizer also supports multi-label classification. 

> How does one determine the appropriate values for n of iterations, compounding factor, and dropout rate?

Empirically. You probably won't need to change the compounding factor, and on most problems dropout of 0.0, 0.1 or 0.2 is good. For the number of epochs, you can just see how long it takes before the model stops improving on development data.

> When training a blank model with 4-5 entities, is it better to train them at once or start with just one? What about the length of sentences/paragraphs to train on?

It's up to you, it depends on whether the entities can overlap, and whether the results will be easier to work with if you do it individually. I would normally default to training them at once if possible, as it's more efficient.

> Is it a common practice to train over and over again on one given model or, rather, train one time once all data is collected? Is consequent training (or even consequent training on the same training data) detrimental to the model?

You should normally re-initialize the model randomly and train as a batch, making several passes over the data, shuffling after each epoch. The `spacy train` command will do all this for you. The results are much easier to reason about this way, because you're treating each example as equally important, and you don't have to think about ordering effects.

The alternative you're suggesting is usually called "online learning", and it's a much more tricky setup that you should only use if constraints force you into it, for instance if there's some reason why retraining is impractical. But even then, it's normally better to just retrain less frequently, e.g. once a week or something.
